### PR TITLE
Make training shuffle a train-only option; migrate config and use samplers for ordering

### PR DESCRIPTION
**Pull request by mtauraso**
### Motivation

- Consolidate dataloader ordering control to a single train-only config key so training sample shuffling is explicit and other verbs remain deterministic.  
- Remove the legacy behavior of passing `shuffle` through PyTorch `DataLoader` kwargs to avoid sampler vs shuffle conflicts and to support explicit subset samplers when `split_indices` are used.  
- Provide a migration path so existing `data_loader.shuffle` entries are preserved and emit deprecation warnings.

### Description

- Add documentation `specs/train_shuffle_config.md` describing that `train.shuffle` is train-only and how it differs from `setup_dataset(..., shuffle=...)`.  
- Add migration `src/hyrax/config_migrations/migrations/002_move_data_loader_shuffle_to_train.py` to rename `data_loader.shuffle` → `train.shuffle`.  
- Update default config `src/hyrax/hyrax_default_config.toml` to include `train.shuffle = true` and remove `data_loader.shuffle`.  
- Change `dist_data_loader` in `src/hyrax/pytorch_ignite.py` to accept a `shuffle` parameter, to warn and ignore legacy `data_loader.shuffle`, to build deterministic `torch.Generator` seeding, and to create explicit `SubsetRandomSampler` or `SubsetSequentialSampler` via a `make_sampler` helper instead of passing `shuffle` into `DataLoader`.  
- Wire `train` verb to read `config['train']['shuffle']` and only apply shuffling to the `train` split; keep validation and test splits deterministic.  
- Remove per-verb code that forcibly mutated `config['data_loader']['shuffle']` for `infer` and `test`.  
- Add and update unit tests to cover migration, default config placement, sampler selection, legacy-key warnings, and runtime behavior across `train`, `infer`, and legacy multi-split paths (tests under `tests/hyrax/*`).

### Testing

- Ran the test suite with `pytest` focused on config migrations, dataloader split/sampler behavior, infer/train/test integration, and config utils; all modified and added tests passed.  
- New/updated tests include `tests/hyrax/test_config_migrations.py`, `tests/hyrax/test_config_utils.py`, `tests/hyrax/test_infer.py`, `tests/hyrax/test_split_fraction_integration.py`, and `tests/hyrax/test_train.py` and they succeeded under the CI test run.  
- Verified that the migration preserves other `data_loader` keys (e.g. `batch_size`) while moving `shuffle` to `train.shuffle` and that legacy `data_loader.shuffle` emits a `UserWarning`/`DeprecationWarning` as appropriate.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a076c24e0308331a8a33ade0091d945)

### CODE REVIEW

File: src/hyrax/pytorch_ignite.py [hunk 1]
```diff
@@ -184,35 +190,50 @@ def dist_data_loader(
...
-        # When a sampler is present, we must force shuffle=False in the kwargs.
-        loader_kwargs = data_loader_kwargs.copy()
-        if sampler is not None:
-            loader_kwargs["shuffle"] = False
+        sampler = make_sampler(indexes, shuffle)
 
-        # Note that when sampler=None, a default sampler is used. The default config
-        # defines shuffle=False, which should prevent any shuffling of of the data.
-        # We expect that this will be the primary use case when running inference.
-        return idist.auto_dataloader(dataset, sampler=sampler, **loader_kwargs), indexes
```
Comment (mtauraso) [inline src/hyrax/pytorch_ignite.py:215]: On second thought, removing this comment seems okay.

File: src/hyrax/pytorch_ignite.py [hunk 2]
```diff
@@ -224,17 +245,13 @@ def dist_data_loader(
...
-
     # Create the indexes for all splits based on config.
     indexes = create_splits(dataset, config)
 
-    # Create samplers and dataloaders for each split we are interested in
-    samplers = {s: SubsetSequentialSampler(indexes[s]) if indexes.get(s) else None for s in split}
+    # Create samplers and dataloaders for each split we are interested in.
+    # In the legacy multi-split path, the train split is the only split that
+    # honors the shuffle option; validation/test remain deterministic.
+    samplers = {s: make_sampler(indexes[s], shuffle and s == "train") for s in split}
```
[the remark this answers is not in the document]
Comment (mtauraso) [reply src/hyrax/pytorch_ignite.py]: Agree with copilot, but doing it myself.

File: specs/train_shuffle_config.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+```toml
+[data_loader]
+shuffle = false
+```
+
+to:
+
+```toml
+[train]
+shuffle = false
```
Comment (drewoldag) [inline specs/train_shuffle_config.md:54]: As discussed, the config migration will move the shuffle key and value, which will lead to correct reproducibility of previous runs, but the default value in `hyrax_default_config.toml` will become `true`.

File: src/hyrax/pytorch_ignite.py [hunk 4]
```diff
@@ -158,6 +158,7 @@ def dist_data_loader(
     dataset: Dataset,
     config: dict,
     split: Union[str, list[str], bool] = False,
```
Comment (drewoldag) [inline src/hyrax/pytorch_ignite.py:160]: FWIW - I think we might be close to being able to remove this argument. It is only non-False in a couple of legacy call points in `train`.

File: src/hyrax/pytorch_ignite.py [hunk 5]
```diff
@@ -184,35 +190,50 @@ def dist_data_loader(
...
-    # Extract the config dictionary that will be provided as kwargs to the DataLoader
+    # Extract the config dictionary that will be provided as kwargs to the DataLoader.
+    # Hyrax controls ordering through explicit samplers; warn and ignore legacy
+    # ``data_loader.shuffle`` if an old/unversioned config still contains it.
     data_loader_kwargs = dict(config["data_loader"])
+    if "shuffle" in data_loader_kwargs:
+        msg = (
+            "config['data_loader']['shuffle'] is ignored and is not passed to PyTorch DataLoader. "
+            "Hyrax controls dataloader ordering with explicit samplers; use config['train']['shuffle'] "
+            "to control training sample shuffling."
```
Comment (drewoldag) [inline src/hyrax/pytorch_ignite.py]:
```suggestion
            "to control training sample shuffling to support reproducibility."
```

File: src/hyrax/pytorch_ignite.py [hunk 6]
```diff
@@ -184,35 +190,50 @@ def dist_data_loader(
...
+        )
+        warnings.warn(msg, UserWarning, stacklevel=2)
```
Comment (drewoldag) [inline src/hyrax/pytorch_ignite.py]: Remove this warning

File: src/hyrax/pytorch_ignite.py [hunk 7]
```diff
@@ -184,35 +190,50 @@ def dist_data_loader(
...
+        logger.warning(msg)
+        data_loader_kwargs.pop("shuffle")
 
     # TODO: Actually DataProvider.collate. Callsites and parameter signature above have not been updated.
```
Comment (drewoldag) [inline src/hyrax/pytorch_ignite.py:206]: Outside of the scope of this PR, but I think this comment can be removed.

File: tests/hyrax/test_config_utils.py [hunk 8]
```diff
@@ -475,3 +475,13 @@ def test_set_config_overwrite():
...
+
+
+def test_default_shuffle_config_is_train_only():
+    """Default shuffle config lives under train and is absent from data_loader."""
+    import hyrax
+
+    config = hyrax.Hyrax().config
+
+    assert config["train"]["shuffle"] is True
+    assert "shuffle" not in config["data_loader"]
```
Comment (drewoldag) [inline tests/hyrax/test_config_utils.py]: This test shouldn't be in here. Not sure that it's really necessary, but if we keep it, I think it should be in test_config_migrations.py. 

This test suite is for the config_utils.py functionality, no for verifying the contents of the config dictionary.

File: tests/hyrax/test_infer.py [hunk 9]
```diff
@@ -49,13 +49,13 @@ def test_infer_trace(loopback_hyrax):
...
-@pytest.mark.parametrize("shuffle", [True, False])
-def test_infer_order(loopback_hyrax, shuffle):
+@pytest.mark.parametrize("train_shuffle", [True, False])
+def test_infer_order(loopback_hyrax, train_shuffle):
     """Test that the order of data run through infer
     is correct in the presence of several splits
     """
     h, dataset = loopback_hyrax
-    h.config["data_loader"]["shuffle"] = shuffle
+    h.config["train"]["shuffle"] = train_shuffle
```
Comment (drewoldag) [inline tests/hyrax/test_infer.py:58]: Yes, I like this very much!

**Review by drewoldag — APPROVED** [review]
This looks good. A few comments, but nothing overly pressing. Remove the `warning`, move or maybe remove one of the tests. 

Otherwise good to go.

File: src/hyrax/pytorch_ignite.py (hunk 6 above)
Comment (mtauraso) [inline src/hyrax/pytorch_ignite.py]:
```suggestion
```

File: tests/hyrax/test_infer.py (hunk 9 above)
Comment (aritraghsh09) [reply tests/hyrax/test_infer.py:58]: +1

**Comment by aritraghsh09** [comment]
Agree with the overall design choice here. Deferring to drew for the actual review

File: tests/hyrax/test_config_utils.py (hunk 8 above)
Comment (mtauraso) [reply tests/hyrax/test_config_utils.py]: I'm going with removal.

Comment (mtauraso) [inline tests/hyrax/test_config_utils.py]:
```suggestion
```

**Comment by mtauraso** [comment]
@copilot Can you change the failing tests so they don't look for the warning that I specifically removed. kthx
